base_layer/core/src/transaction_protocol/receiver.rs

```diff
@@ -0,0 +1,63 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    transaction::TransactionOutput,
+    transaction_protocol::TransactionProtocolError,
+    types::{PublicKey, SecretKey, Signature},
+};
+use std::collections::HashMap;
+use tari_crypto::challenge::MessageHash;
+
+pub enum RecipientState {
+    Finalized(RecipientSignedTransactionData),
+    Failed(TransactionProtocolError),
+}
+
+/// An enum describing the types of information that a recipient can send back to the receiver
+#[derive(Debug, Clone)]
+pub(super) enum RecipientInfo {
+    None,
+    Single(Option<Box<RecipientSignedTransactionData>>),
```

Comment:
Yes. An enum is sized as `max` size of every field. So without a Box this would take up `RecipientSignedTransactionData` size of memory, even if the value is None, or Single(None). With a Box, the comiler allocates 8bytes only.

Reply:
Interesting!